File: 2015/15/solution.php
<?php

class Ingredient
{
    public $capacity;
    public $durability;
    public $flavor;
    public $texture;
    public $calories;

    public function __construct(int $capacity, int $durability, int $flavor, int $texture, int $calories)
    {
        $this->capacity   = $capacity;
        $this->durability = $durability;
        $this->flavor     = $flavor;
        $this->texture    = $texture;
        $this->calories   = $calories;
    }
}

function best_recipe_score(array &$ingredients, int $diet = null, int $size = 100, int $depth = 0, array $amounts = [])
{
    switch ($depth - count($ingredients)) {
        case 0:
            if (array_sum($amounts) !== $size) {
                return PHP_INT_MIN;
            }
            $capacity = $durability = $flavor = $texture = $calories = 0;
            for ($i = 0; $i < count($ingredients); $i++) {
                $capacity   += $amounts[$i] * $ingredients[$i]->capacity;
                $durability += $amounts[$i] * $ingredients[$i]->durability;
                $flavor     += $amounts[$i] * $ingredients[$i]->flavor;
                $texture    += $amounts[$i] * $ingredients[$i]->texture;
                $calories   += $amounts[$i] * $ingredients[$i]->calories;
            }
            if (!is_null($diet) && $diet !== $calories) {
                return PHP_INT_MIN;
            }

            return max(0, $capacity) * max(0, $durability) * max(0, $flavor) * max(0, $texture);

        case 1:
            return best_recipe_score(
                $ingredients,
                $diet,
                $size,
                $depth + 1,
                array_merge($amounts, [$depth => $size - array_sum($amounts)])
            );

        default:
            $max = PHP_INT_MIN;
            for ($amount = 100 - array_sum($amounts); $amount >= 0; $amount--) {
                $max = max($max, best_recipe_score(
                    $ingredients,
                    $diet,
                    $size,
                    $depth + 1,
                    array_merge($amounts, [$depth => $amount])
                ));
            }

            return $max;
    }
}

$ingredients = [];
while (preg_match('/\w+: capacity (-?\d+), durability (-?\d+), flavor (-?\d+), texture (-?\d+), calories (-?\d+)/', fgets(STDIN), $matches)) {
    $ingredients[] = new Ingredient($matches[1], $matches[2], $matches[3], $matches[4], $matches[5]);
}

fwrite(STDOUT, best_recipe_score($ingredients) . PHP_EOL);
fwrite(STDOUT, best_recipe_score($ingredients, 500) . PHP_EOL);

File: 2015/15/solution.java

import java.util.ArrayList;
import java.util.Scanner;

public class AoC_15 {

    class Ingredient {

        long capacity, durability, flavor, texture, calories, usage;

        Ingredient(String[] input) {
            capacity = Integer.valueOf(input[2].substring(0, input[2].length() - 1));
            durability = Integer.valueOf(input[4].substring(0, input[4].length() - 1));
            flavor = Integer.valueOf(input[6].substring(0, input[6].length() - 1));
            texture = Integer.valueOf(input[8].substring(0, input[8].length() - 1));
            calories = Integer.valueOf(input[10]);
        }
    }

    void magic() {
        ArrayList<Ingredient> ingredients = new ArrayList<>();
        while (sc.hasNextLine()) {
            ingredients.add(new Ingredient(sc.nextLine().split(" ")));
        }
        System.out.println("Part One: " + recurse(ingredients, 0, 0, false));
        System.out.println("Part Two: " + recurse(ingredients, 0, 0, true));
    }

    long recurse(ArrayList<Ingredient> ingredients, int used, int depth, boolean limitCalories) {
        if (depth == ingredients.size()) {
            long capacity = 0, durability = 0, flavor = 0, texture = 0, calories = 0;
            for (Ingredient ing : ingredients) {
                capacity += ing.capacity * ing.usage;
                durability += ing.durability * ing.usage;
                flavor += ing.flavor * ing.usage;
                texture += ing.texture * ing.usage;
                calories += ing.calories * ing.usage;
            }
            if (used != 100 || capacity < 0 || durability < 0 || flavor < 0 || texture < 0 || (limitCalories && calories != 500l)) {
                return 0l;
            }
            return capacity * durability * flavor * texture;
        }
        long max = 0l;
        for (int i = 100 - used; i >= 0; i--) {
            ingredients.get(depth).usage = i;
            max = Math.max(max, recurse(ingredients, used + i, depth + 1, limitCalories));
        }
        return max;
    }

    static Scanner sc = new Scanner(System.in);

    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();
        new AoC_15().magic();
        System.out.println("Running time: " + (System.currentTimeMillis() - startTime) + "ms");
    }
}
